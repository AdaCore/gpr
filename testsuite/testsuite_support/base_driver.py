import os
import os.path
from typing import NoReturn

from e3.env import Env
from e3.fs import mkdir
from e3.testsuite.driver.classic import TestAbortWithError
from e3.testsuite.driver.diff import DiffTestDriver, ReplacePath, Substitute


# create_fake_ada_compiler routine copied from gprbuild-internal testsuite
# support code.


def create_fake_ada_compiler(
    driver,
    comp_dir,
    comp_target,
    gnat_version,
    gcc_version,
    comp_is_cross=False,
    runtimes=["native", "sjlj"],
    create_symlink=False,
    create_ada_object_path=False,
):
    """
    Create directory defined by the comp_dir parameter and put fake Ada
    compiler directory tree there. If comp_is_cross is true, the compiler
    tools 'gnatmake', 'gcc', and 'gnatls' will be prefixed by the
    comp_target. If create_symlink is true, the first runtime from the
    runtimes will be made available as default through an 'adalib' symbolic
    link.
    If create_ada_object_path is true, that file will be created to simulate
    a Windows install.
    """

    if comp_is_cross:
        comp_prefix = comp_target + "-"
    else:
        comp_prefix = ""

    env = Env()
    comp_dict = {
        "comp_target": comp_target,
        "gnat_version": gnat_version,
        "gcc_version": gcc_version,
        "comp_prefix": comp_prefix,
        "exeext": env.build.os.exeext,
    }

    mkdir(os.path.join(comp_dir, "bin"))
    gnatls_adb = open(os.path.join(comp_dir, "bin", "gnatls.adb"), "w")
    gnatls_adb.write(
        """
with Ada.Text_IO; use Ada.Text_IO;
with Ada.Command_Line; use Ada.Command_Line;
procedure gnatls is
begin
   if Argument_Count >= 1 and Argument (1) = "-v" then
        Put_Line ("GNATLS Pro %(gnat_version)s (20190507-89)");
   else
         Put ("Running gnatls");
         for J in 1 .. Argument_Count loop
             Put (" " & Argument (J));
         end loop;
   end if;
end gnatls;
"""
        % comp_dict
    )
    gnatls_adb.close()

    gcc_adb = open(os.path.join(comp_dir, "bin", "gcc.adb"), "w")
    gcc_adb.write(
        """
with Ada.Text_IO; use Ada.Text_IO;
with Ada.Command_Line; use Ada.Command_Line;
procedure gcc is
begin
   if Argument_Count >= 1 and then Argument (1) = "-v" then
        Put_Line ("gcc version %(gcc_version)s 20131008 for GNAT Pro");
   elsif Argument_Count >= 1 and then Argument (1) = "--version" then
        Put_Line ("gcc (GCC) %(gcc_version)s");
   elsif Argument_Count >= 1 and then Argument (1) = "-dumpmachine" then
        Put_Line ("%(comp_target)s");
   else
         Put ("Running gcc");
         for J in 1 .. Argument_Count loop
             Put (" " & Argument (J));
         end loop;
   end if;
end gcc;
"""
        % comp_dict
    )
    gcc_adb.close()

    gnatmake_adb = open(os.path.join(comp_dir, "bin", "gnatmake.adb"), "w")
    gnatmake_adb.write(
        """
with Ada.Text_IO; use Ada.Text_IO;
with Ada.Command_Line; use Ada.Command_Line;
procedure gnatmake is
begin
         Put ("Running gcc");
         for J in 1 .. Argument_Count loop
             Put (" " & Argument (J));
         end loop;
end gnatmake;
"""
    )
    gnatmake_adb.close()

    for tool in ["gnatmake", "gcc", "gnatls"]:
        comp_dict["bin"] = tool

        # Do not run gnatmake in the same directory with the fake tools sources
        # to avoid using just created fake tools in the build process.

        driver.shell(
            [
                "gnatmake",
                os.path.join("bin", tool + ".adb"),
                "-o",
                os.path.join("bin", "%(comp_prefix)s%(bin)s%(exeext)s" % comp_dict),
            ],
            cwd=comp_dir,
            analyze_output=False,
        )

    if comp_target == "dotnet":
        for dir in ("adalib", "adainclude"):
            mkdir(os.path.join(comp_dir, "lib", "dotgnat", dir))
    else:
        for runtime in runtimes:
            for dir in ("adalib", "adainclude"):
                mkdir(
                    os.path.join(
                        comp_dir,
                        "lib",
                        "gcc",
                        comp_target,
                        gcc_version,
                        "rts-%s" % runtime,
                        dir,
                    )
                )

    libdir = os.path.join(comp_dir, "lib", "gcc", comp_target, gcc_version)

    # On Unix systems, we have a symbolic link for the default
    # runtime. gprconfig should automatically detect these are
    # the same two runtimes and only list "native".

    if create_symlink:
        os.symlink(
            os.path.join("rts-%s" % runtimes[0], "adalib"),
            os.path.join(libdir, "adalib"),
        )

    # Simulate windows system, with an ada_object_path file

    if create_ada_object_path:
        with open(os.path.join(libdir, "ada_object_path"), "w") as ada_obj:
            ada_obj.write("rts-%s/adalib" % runtimes[0])


class BaseDriver(DiffTestDriver):
    """Base class to provide common test driver helpers."""

    def set_up(self):
        super(BaseDriver, self).set_up()

        description = self.test_env.get("description", None)
        if description is None:
            raise TestAbortWithError('test.yaml: missing "description" field')
        output = self.test_env.get("output", None)
        if output is not None:
            test_out = None
            if not isinstance(output, list):
                raise TestAbortWithError('test.yaml: "output" field must be a list')

            for i, entry in enumerate(output, 1):

                def error(message: str) -> NoReturn:
                    raise TestAbortWithError(
                        "test.yaml: output.entry#{}: {}".format(i, message)
                    )

                if (
                    not isinstance(entry, list)
                    or len(entry) != 2
                    or any(not isinstance(s, str) for s in entry)
                ):
                    error("list of 2 strings expected")
                try:
                    cond_env = dict()
                    cond_env["env"] = self.env
                    cond = eval(entry[1], cond_env)
                except Exception as exc:
                    error("invalid condition ({}): {}".format(type(exc).__name__, exc))
                if cond:
                    if test_out is not None:
                        error(
                            f"several test.out selected: {test_out} and {entry[0]}"
                        )
                    test_out = entry[0]
            if test_out is not None:
                self.test_env["baseline_file"] = test_out

    @property
    def output_refiners(self):
        # Remove working directory from output and
        # make all filenames look like Unix ones (forward slashes for directory
        # separators, no drive letter).
        return super().output_refiners + [
            ReplacePath(self.working_dir(), replacement=""),
            Substitute("\\", "/"),
            Substitute("C:/", "/"),
            Substitute("x86_64-pc-linux-gnu", replacement="(host)"),
            Substitute("i686-pc-linux-gnu", replacement="(host)"),
            Substitute("x86_64-w64-mingw32", replacement="(host)"),
            Substitute("i686-pc-mingw32", replacement="(host)"),
            Substitute("x86_64-linux", replacement="(host)"),
            Substitute("x86-linux", replacement="(host)"),
            Substitute("x86_64-windows", replacement="(host)"),
            Substitute("x86-windows", replacement="(host)"),
        ]

    # Convenience path builders

    @property
    def testsuite_dir(self):
        """Return the absolute path to the testsuite root directory."""
        result = os.path.join(os.path.dirname(os.path.abspath(__file__)), "..")
        return os.path.abspath(result)
