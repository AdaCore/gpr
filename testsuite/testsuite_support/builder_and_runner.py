import os
from e3.os.process import PIPE, Run, STDOUT
from random import getrandbits
from e3.testsuite.driver.classic import TestAbortWithFailure

# environment variables definition

USE_VALGRIND = "USE_VALGRIND"
USE_GNATCOV = "USE_GNATCOV"
COV_TRACES_DIR = "GNATCOV_TRACES_DIR"
COV_LEVEL = "GNATCOV_LEVEL"

GPR = os.environ["GPR2_TOOLS_PREFIX"] if "GPR2_TOOLS_PREFIX" in os.environ else "gpr"
GPRLS = GPR + "ls"
GPRCLEAN = GPR + "clean"
GPRINSTALL = GPR + "install"
GPRNAME = GPR + "name"
GPRCONFIG = GPR + "config"
GPRREMOTE = GPR + "remote"
GPRDOC = GPR + "doc"
GPRINSPECT = GPR + "inspect"


class BuilderAndRunner(object):
    """
    Handle program using gpr2 build. in coverage mode, instrument code &
    build instrumented code.

    Run program using gpr2 (including prebuilt gpr2-tools) using valgrind when
    requested.

    When using BuilderAndRunner() object is configured using python script
    parameters --valgrind --gnatcov --level --traces-dir

    In coverage mode, gpr2 programs's coverage traces are written in
    traces-dir using a random filename.

    This object provides wrappers for e3.os.process.Run, subprocess.call &
    subprocess.check_output
    """

    def __init__(self, driver=None):
        """configure BuilderAndRunner object using TestDriver environment or
        python script's argument when driver is None.
        """

        # associated TestDriver object
        self.driver = driver

        if driver is not None:

            # valgrind mode enabled status
            self.valgrind = driver.env.valgrind

            # coverage mode enabled status
            self.gnatcov = driver.env.gnatcov is not None

            if self.gnatcov:
                # gnatcov trace files directory
                self.traces_dir = driver.env.gnatcov.traces_dir

                # gnatcov calls --level switch
                self.level = driver.env.gnatcov.covlevel
            else:
                self.level = None
                self.traces_dir = None
        else:
            self.valgrind = USE_VALGRIND in os.environ
            self.gnatcov = USE_GNATCOV in os.environ
            self.traces_dir = os.environ.get(COV_TRACES_DIR)
            self.level = os.environ.get(COV_LEVEL)

    def simple_run(
        self,
        cmd,
        env=None,
        catch_error=True,
        output=PIPE,
        error=STDOUT,
        analyze_output=True,
    ):
        """ generic TestDriver.shell or e3.os.process.Run runner"""
        if self.driver is not None:
            if env is not None:
                effective_env = dict(os.environ)
                effective_env.update(env)
            return self.driver.shell(
                cmd,
                env=effective_env,
                catch_error=catch_error,
                analyze_output=analyze_output,
            )
        else:
            p = Run(cmd, env=env, output=output, error=error, ignore_environ=False)
            if catch_error and p.status != 0:
                print(str(cmd) + " returned " + str(p.status))
                if p.out:
                    print("stdout\n" + p.out)
                if p.err:
                    print("stderr\n" + p.err)
                raise TestAbortWithFailure("non-zero exit status")
            else:
                return p

    def build(self, project, vars=[], args=[], env=None, output=PIPE):
        """ gprbuild wrapper for normal & coverage modes """
        # If code coverage is requested, leave a chance to gnatcov to decorate
        # the execution of the subprogram in order to make it contribute to
        # code coverage.
        if self.gnatcov:
            gnatcov_cmd = [
                "gnatcov",
                "instrument",
                "--level",
                self.level,
                "--dump-trigger=atexit",
                "--externally-built-projects",
                "--projects",
                "gpr2",
                "--no-subprojects",
                "-P",
                project,
            ] + vars
            self.simple_run(gnatcov_cmd, env=env, analyze_output=False)

            # to do coverage, we use the repo's gpr2 project file instead of
            # the installed one from libgpr2. This means we need to ensure
            # that some scenario variables are properly set.
            gprbuild_cmd = (
                ["gprbuild", "-P", project,
                 "-XGPR2_BUILD=gnatcov", "-XXMLADA_BUILD=static"]
                + vars
                + ["--src-subdirs=gnatcov-instr", "--implicit-with=gnatcov_rts"]
                + args
            )
        else:
            gprbuild_cmd = ["gprbuild", "-P", project] + vars + args

        return self.simple_run(gprbuild_cmd, env=env, output=output)

    def run(self, cmd, env=None, output=PIPE, catch_error=False):
        """ generic wrapper handling coverage & valgrind modes. """
        if self.gnatcov:
            if env is None:
                env = {}
            file = str(getrandbits(128)) + ".srctrace"
            env["GNATCOV_TRACE_FILE"] = os.path.join(self.traces_dir, file)
        run_cmd = (["valgrind", "-q"] if self.valgrind else []) + cmd

        return self.simple_run(run_cmd, env=env, catch_error=catch_error, output=output)

    def check_output(self, cmd):
        """subprocess.check_output wrapper handling coverage & valgrind
        modes.
        """
        return self.run(cmd, catch_error=True)

    def call(self, cmd):
        """subprocess.call wrapper handling coverage & valgrind
        modes.
        """
        p = self.run(cmd)
        print(p.out, end="")
        return p

    def check_call(self, cmd):
        """subprocess.call wrapper handling coverage & valgrind
        modes.
        """
        p = self.run(cmd, catch_error=True)
        print(p.out, end="")
        return p

    def insert_build_and_runner_parameters(self, env):
        """The BuildAndRunner initialization parameters to allow a
        BuilderAndRunner() call in the python test script.
        """
        if self.valgrind:
            env[USE_VALGRIND] = "true"
        if self.gnatcov:
            env[USE_GNATCOV] = "true"
            env[COV_TRACES_DIR] = self.traces_dir
            env[COV_LEVEL] = self.level
