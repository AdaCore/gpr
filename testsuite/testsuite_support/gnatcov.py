import glob
import logging
import os.path
import shutil

from e3.os.process import Run, quote_arg


class GNATcov(object):
    """
    Helper to compute code coverage with GNATcoverage using the testsuite.

    After initialization, the workflow for this helper is:

    * call the `report()` method to generate coverage reports.
    """

    def __init__(self, testsuite):
        """
        :param Testsuite testsuite: Testsuite instance to work with.
        """
        self.level = testsuite.env.options.level
        self.sid_dir = testsuite.env.options.gnatcov

        self.temp_dir = os.path.join(testsuite.env.working_dir, 'gnatcov')
        self.output_dir = testsuite.output_dir

        self.ensure_clean_dir(self.temp_dir)
        os.mkdir(self.traces_dir)

    @property
    def traces_dir(self):
        """ The gnatcov's trace file destination folder """
        return os.path.join(self.temp_dir, 'traces')

    @staticmethod
    def ensure_clean_dir(dirname):
        """
        If it exists, remove the ``dirname`` directory tree and create an empty
        directory instead.
        """
        if os.path.exists(dirname):
            shutil.rmtree(dirname)
        os.mkdir(dirname)

    @staticmethod
    def checked_run(argv):
        """
        Run a process with the given arguments. Log its output and raise an
        error if it fails.
        """
        p = Run(argv)
        if p.status != 0:
            logging.error('Command failed: %s',
                          ' '.join(quote_arg(arg) for arg in argv))
            logging.error('Output:\n' + p.out)
            raise RuntimeError

    def report(self, formats=['dhtml', 'xcov']):
        """Generate coverage reports for all given output formats."""

        # Get the list of all trace files
        traces_list = os.path.join(self.temp_dir, 'traces.txt')
        with open(traces_list, 'w') as f:
            for t in glob.glob(os.path.join(self.traces_dir, '*.srctrace')):
                f.write(t + '\n')

        # Load trace files only once, produce a checkpoint for them
        logging.info('Consolidating coverage results')
        ckpt_file = os.path.join(self.temp_dir, 'report.ckpt')
        self.checked_run(['gnatcov', 'coverage', '--level', self.covlevel,
                          '-P', 'gpr2-tools',
                          '-XGPR2_BUILD=gnatcov',
                          '--externally-built-projects',
                          '--save-checkpoint', ckpt_file,
                          '@' + traces_list])

        # Now, generate all requested reports from this checkpoint
        logging.info('Generating coverage reports ({})'
                     .format(', '.join(sorted(formats))))
        for fmt in formats:
            report_dir = os.path.join(self.output_dir, 'coverage-' + fmt)
            self.ensure_clean_dir(report_dir)
            self.checked_run([
                'gnatcov', 'coverage',
                '--annotate', fmt,
                '--level', self.covlevel,
                '--output-dir', report_dir,
                '-P', 'gpr2-tools',
                '--externally-built-projects',
                '-XGPR2_BUILD=gnatcov',
                '--checkpoint', ckpt_file])

    @property
    def covlevel(self):
        """ The gnatcov's --level value used during this testsuite's run """
        return self.level
