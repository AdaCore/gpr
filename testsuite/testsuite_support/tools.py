import os

# Environment variable names for each tool
GPRLS_NAME = "GPRLS_NAME"
GPRCLEAN_NAME = "GPRCLEAN_NAME"
GPRINSTALL_NAME = "GPRINSTALL_NAME"
GPRCONFIG_NAME = "GPRCONFIG_NAME"
GPRDOC_NAME = "GPRDOC_NAME"
GPRINSPECT_NAME = "GPRINSPECT_NAME"
GPRBUILD_NAME = "GPRBUILD_NAME"

GPRLS = os.environ[GPRLS_NAME] if GPRLS_NAME in os.environ else "gprls"
GPRCLEAN =  os.environ[GPRCLEAN_NAME] if GPRCLEAN_NAME in os.environ else "gprclean"
GPRINSTALL =  os.environ[GPRINSTALL_NAME] if GPRINSTALL_NAME in os.environ else "gprinstall"
GPRCONFIG =  os.environ[GPRCONFIG_NAME] if GPRCONFIG_NAME in os.environ else "gprconfig"
GPRDOC =  os.environ[GPRDOC_NAME] if GPRDOC_NAME in os.environ else "gprdoc"
GPRINSPECT =  os.environ[GPRINSPECT_NAME] if GPRINSPECT_NAME in os.environ else "gprinspect"
GPRBUILD = os.environ[GPRBUILD_NAME] if GPRBUILD_NAME in os.environ else "gprbuild"
