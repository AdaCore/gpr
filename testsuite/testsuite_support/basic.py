import os
import re
from e3.testsuite.driver.classic import TestAbortWithError
from e3.testsuite.driver.classic import ClassicTestDriver
from e3.fs import cp
from testsuite_support.base_driver import create_fake_ada_compiler
from testsuite_support.builder_and_runner import BuilderAndRunner

TESTSUITE_ROOT_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
ADA_SUPPORT_FILES = os.path.join(TESTSUITE_ROOT_DIR, "support")


class BasicDriver(ClassicTestDriver):
    """
    Driver to build a test program and run it.

    Interface:

    * put a main in test.adb that will be compiled (using a default project)
    * use Test_Assert functions to make your assertions
    * return Test_Assert.Report at the end of our main procedure

    This driver will build the program using GPRbuild and will then run the
    program and ensure that all assertions did passed and that the program
    did not crashed
    """

    def set_up(self):
        self.project_file = self.test_env.get("project_file")
        if self.project_file is None:
            cp(
                os.path.join(ADA_SUPPORT_FILES, "test.gpr"),
                self.test_env["working_dir"],
            )
            self.project_file = "./test.gpr"
            self.main = "./obj/test"
        else:
            self.main = self.test_env.get("./main")

        if not self.project_file or not isinstance(self.project_file, str):
            raise TestAbortWithError(
                'test.yaml: please define a "project_file" string field'
            )
        if not self.main or not isinstance(self.main, str):
            raise TestAbortWithError('test.yaml: please define a "main" string field')

        self.fake_ada_target = self.test_env.get("fake_ada_target")
        self.builder_and_runner = BuilderAndRunner(self)

    def run(self):
        env = {"ADA_SUPPORT_SOURCE_DIR": ADA_SUPPORT_FILES}

        # If we are requested to run with a fake toolchain, set it up now
        if self.fake_ada_target:
            fake_dir = self.working_dir("fake-ada")
            create_fake_ada_compiler(
                self,
                comp_dir=fake_dir,
                comp_target=self.fake_ada_target,
                gnat_version="21.0w",
                gcc_version="8.4.3",
                runtimes=["rtp"],
                comp_is_cross=True,
            )
            env["PATH"] = os.path.join(fake_dir, "bin") + os.path.pathsep + env["PATH"]

        # Build the program and run it
        p = self.builder_and_runner.build(
            project=self.project_file,
            args=["-g1", "-p", "-bargs", "-Es"],
            env=env,
        )
        self.output += p.out
        p = self.builder_and_runner.run([self.main], env=env)
        self.output += p.out
        if "Call stack traceback locations" in p.out:
            m = re.search(r"Call stack traceback locations:.*\n([^\r\n]+)", p.out)
            if m.group(1) is not None:
                self.shell(["addr2line", "-e", self.main] + m.group(1).split())

    def compute_failures(self):
        return (
            ["Success marker not found"]
            if "<=== TEST PASSED ===>" not in self.result.log.log
            else []
        )
