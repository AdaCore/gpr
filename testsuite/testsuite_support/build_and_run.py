import os
import os.path

from e3.testsuite.driver.classic import TestAbortWithError
from e3.env import Env

from testsuite_support.base_driver import BaseDriver, create_fake_ada_compiler
from testsuite_support.builder_and_runner import BuilderAndRunner


class BuildAndRunDriver(BaseDriver):
    """
    Driver to build a test program and run it.

    Interface:

    * put a project file for a program and the associated source files in the
      test directory;
    * put a "test.out" text file in the test directory;
    * in the "test.yaml" file, add a "project_file" key that contains the name
      of the project file, and add also a "main" key that contains the name of
      the program to run.

    This driver will build the program using GPRbuild and will then run the
    program. The output of this run is checked against the expected output
    (test.out file).
    """

    def set_up(self):
        super(BuildAndRunDriver, self).set_up()

        project_file = self.test_env.get("project_file", None)
        main = self.test_env.get("main", None)

        if not project_file or not isinstance(project_file, str):
            raise TestAbortWithError(
                'test.yaml: please define a "project_file" string field'
            )
        if not main or not isinstance(main, str):
            raise TestAbortWithError('test.yaml: please define a "main" string field')

        self.project_file = project_file
        self.main_program = main
        self.fake_ada_target = self.test_env.get("fake_ada_target", None)

        self.builder_and_runner = BuilderAndRunner(self)

    def run(self):
        env = {}

        # Build the program and run it
        # Note: do it *before* creating fake toolchains, in case a native
        # fake compiler is requested.
        self.builder_and_runner.build(
            project=self.project_file,
            args=["-g1", "-q", "-p", "-bargs", "-Es"],
            env=env,
        )

        # If we are requested to run with a fake toolchain, set it up now
        if self.fake_ada_target:
            if isinstance(self.fake_ada_target, list):
                targets = self.fake_ada_target
            else:
                targets = [self.fake_ada_target]
            paths = []
            for tgt in targets:
                fake_dir = self.working_dir("fake-ada-%s" % tgt)
                paths.append(os.path.join(fake_dir, "bin"))

                if 'linux' in tgt or 'windows' in tgt:
                    is_cross = False
                elif tgt == 'native':
                    is_cross = False
                    tgt = Env().host.triplet
                else:
                    is_cross = True
                if not is_cross:
                    rts = ["native", "sjlj", "light"]
                elif 'vxworks' in tgt:
                    rts = ["rtp"]
                else:
                    assert False, "unexpected fake target %s" % tgt

                create_fake_ada_compiler(
                    self,
                    comp_dir=fake_dir,
                    comp_target=tgt,
                    gnat_version="21.0w",
                    gcc_version="8.4.3",
                    runtimes=rts,
                    comp_is_cross=is_cross,
                )
            paths.append(os.environ.get("PATH"))
            env["PATH"] = os.pathsep.join(paths)

        self.builder_and_runner.run([os.path.join(".", self.main_program)], env=env)
