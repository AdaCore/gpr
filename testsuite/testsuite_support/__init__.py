import e3.testsuite

from .build_and_run import BuildAndRunDriver
from .basic import BasicDriver
from testsuite_support.gnatcov import GNATcov
from testsuite_support.python_script import PythonScriptDriver


class Testsuite(e3.testsuite.Testsuite):
    tests_subdir = 'tests'
    enable_cross_support = True
    test_driver_map = {
        'build_and_run': BuildAndRunDriver,
        'python_script': PythonScriptDriver,
        'basic': BasicDriver
    }

    @property
    def auto_generate_text_report(self) -> bool:
        "Generate a text report together with the raw results file"
        return True

    def add_options(self, parser):
        parser.add_argument(
            "--valgrind",
            action="store_true",
            help="Run test executable under valgrind.")

        parser.add_argument(
            "--from-gnat",
            action="store_true",
            help="If provided, means gnat packages are available")

        parser.add_argument(
            "--gnatcov",
            help="If provided, compute the source code coverage of testcases"
                 " on ALS. This requires GNATcoverage working with"
                 " instrumentation. The argument passed must be a directory"
                 " that contains all SID files.")

        parser.add_argument(
            "--level",
            help="coverage level. can be:"
                 " branch, insn, stmt, stmt+decision, stmt+mcdc, stmt+uc_mcdc")

    def set_up(self):
        super(Testsuite, self).set_up()
        self.env.valgrind = self.main.args.valgrind
        self.env.from_gnat = self.main.args.from_gnat

        # If code coverage is requested, initialize our helper and build
        # instrumented programs.
        self.env.gnatcov = (GNATcov(self) if self.main.args.gnatcov else None)
        self.env.root_dir = self.root_dir

    def tear_down(self):
        if self.env.gnatcov:
            self.env.gnatcov.report()

        super(Testsuite, self).tear_down()
