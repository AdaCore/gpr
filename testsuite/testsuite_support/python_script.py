import os
import sys

from testsuite_support.base_driver import BaseDriver
from testsuite_support.builder_and_runner import BuilderAndRunner


class PythonScriptDriver(BaseDriver):
    """
    Driver to run a Python script.

    Interface:

    * put a "test.py" script in the test directory;
    * put a "test.out" text file in the test directory.

    This driver will run the Python script. Its output is then checked against
    the expected output (test.out file). This mechanism is the most flexible
    way to write a testcase, but also the more verbose one and the most complex
    one. Use this driver when no other one fits.
    """

    def run(self):
        builder_and_runner = BuilderAndRunner(self)

        cmd = [sys.executable, 'test.py']

        env = dict(os.environ)
        if 'PYTHONPATH' not in os.environ:
            env['PYTHONPATH'] = self.env.root_dir
        else:
            env['PYTHONPATH'] = str(self.env.root_dir) + \
                                os.path.pathsep + os.environ['PYTHONPATH']
        builder_and_runner.insert_build_and_runner_parameters(env)

        env["root_dir"] = self.env.root_dir

        self.shell(cmd, env=env)
