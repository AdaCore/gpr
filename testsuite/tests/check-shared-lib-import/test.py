from testsuite_support.builder_and_runner import BuilderAndRunner, GPRCLEAN

subprocess = BuilderAndRunner()

subprocess.call([GPRCLEAN, "-r", "prj.gpr"])
subprocess.call([GPRCLEAN, "-r", "--unchecked-shared-lib-imports",
                 "prj.gpr"])

# TODO: reactivate code below when we know the name of switch disabling the
# warnings.
if False:
    subprocess.call([GPRCLEAN, "-ws", "-r",
                     "--unchecked-shared-lib-imports", "prj.gpr"])
