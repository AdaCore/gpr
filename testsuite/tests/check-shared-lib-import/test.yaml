description: Check shared library imports
driver: python_script
