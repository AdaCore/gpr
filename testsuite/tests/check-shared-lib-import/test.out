static.gpr:3:25: warning: library directory "lib" not found
base.gpr:2:6: shared library project "Shared" cannot import project "Nolib" that is not a shared library project
base.gpr:1:6: shared library project "Shared" cannot import static library project "Static"
shared.gpr:5:25: warning: library directory "lib" not found
gprclean: "prj.gpr" processing failed
static.gpr:3:25: warning: library directory "lib" not found
shared.gpr:5:25: warning: library directory "lib" not found
